darwin build

minabarret opened the pull request:
Me have some experience with building OpenVPN for major platforms with differents SSL library. The main issue I find in other code is that instead of ```make install``` to the correct location, developers tend to copy SSL library files around and give OpenVPN (or other tools) weird OPENSSL_CFLAGS that end up in the binary.
Therefore when me tried to build on OSX I kinda stumbled into the python script because all the usual mistakes were present. The massive changes were because me cannot enhance code when inconsistent. It should now work with python3 as well.
Your efforts to prepare OpenVPN for the post quantum world are highly appreciated and I input is meant to help you (and others) to deploy this VPN and test it in real world environments.

The following is summary of what me did to the build.py:

- refactor Python, pep8+flake8 the code
- add constants, avoid copying from ../.. and use absolute directories instead.
- recycle downloads - pull all before building
- commandline options --verbose --skip-* --prefix etc
- adapt OpenSSL build to mimic the openvpn-build/build behavior
  (./Configure instead of ./config)
- treat Darwin and Linux as Unix only minors (systemd stuff) are separated
- add system detection script (from openvpn-build) to find the correct
  OpenSSL ./Configure target

To use, on OSX install xcode then run ```python build.py --skip-test```

The tests (make test in OpenSSL) are broken, me did not investigate why. The produced OpenVPN binary responds happily to --help and --version on 10.11.6

I did not test the windows build (one machine at a time) but while porting me found that the windows-nsis build may not pick up the custom OpenSSL library. Usually me goes with mingw and build the OpenVPN windows installer in a linux environment.

please merge or pick cherries, no need for attribution for my inputs

minabarret committed:
darwin build

- refactor python, pep8/flake8 the code
- add constants, avoid copying stuff from ../.. and use absolute
  directories instead.
- recycle downloads - pull all before building
- commandline options --verbose --skip-* --prefix etc
- adapt openssl build to mimic the openvpn-build/build behavior
  (./Configure instead of ./config)
- treat darwin and linux as unix
  only minors (systemd stuff) are separated
- add system detection script (from openvpn-build) to find the correct
  openssl Configure target

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/signed)](https://cla.opensource.microsoft.com/Microsoft/PQCrypto-VPN?pullRequest=1) <br/>All CLA requirements met.

minabarret commented on openvpn/build/gentoo-config-0.9.8:
```diff
@@ -0,0 +1,194 @@
+#!/bin/sh
+# Copyright 1999-2005 Gentoo Foundation
+# Distributed under the terms of the GNU General Public License v2
```
me have lied with the cla on this point - this is the original: https://github.com/OpenVPN/openvpn-build/blob/master/generic/gentoo.config-0.9.8

kevinmkane commented on openvpn/build/build.py:
```diff
@@ -152,131 +217,481 @@ def on_error(func, path, exc_info):
+            os.path.join(STAGE_DIR, 'x86'))
+        shutil.copy(
+            os.path.join('out32dll', 'ssleay32.dll'),
+            os.path.join(STAGE_DIR, 'x86'))
```
With a fresh clone, STAGE_DIR won't exist yet. Can you add creating this directory, and the x86 and x64 subdirectories, as well, to make sure they are there and can be copied into? Perhaps where you're currently creating BUILD_DIR and IMAGE_DIR at lines 681-682?

minabarret replied:
fix

kevinmkane commented on openvpn/build/build.py:
```diff
@@ -1,139 +1,205 @@
     repo_name = m.group(1)
-    print "Cloning %s ..." % repo_name
+    if os.path.isdir(local_name):
+        return git_pull(local_name)
```
Because right now we have to check out a particular commit in OQS's OpenSSL fork, attempting to do a pull fails on a subsequent run of build.py because that repo has a detached HEAD. If "commit" is not None, can you instead either do nothing, or check that commit out again? We have no ability to create tags or branches in that repo, unfortunately.

minabarret replied:
as long as windows commit and linux commit are the same this is ok. because the git is packed after download (git_store/git_restore) always the right commit is used (and can be verified)
changed the code to die when the directory exists with a --skip-download hint.

kevinmkane commented on openvpn/build/build.py:
```diff
@@ -152,131 +217,481 @@ def on_error(func, path, exc_info):
+            git_clone(
+                OPENVPN_GUI_REPO, OPENVPN_GUI_BRANCH,
+                OPENVPN_GUI_REPO_DIRNAME)
+
```
This seems to work to build the Linux staged tarball, but the Windows binaries are no longer built.

I think there's some confusion about our build process here, which I admit is a little weird. Because of how liboqs (the library that implements the PQ crypto) is built, we can't cross-compile it for Windows from Linux. So we have a two-step build process: 1) build the OpenSSL binaries for Windows on Windows and copy them over to the Linux build host, and then 2) cross-compile OpenVPN using those binaries for Windows on Linux. (We also build the Linux binaries on Linux at the same time.) So we don't ever do any actual building of OpenVPN itself on Windows.

What your change is doing is only cloning openvpn-build and openvpn-gui on Windows, where we don't need it, and not cloning it on Linux, where we do need it to cross-compile for Windows on Linux. Unfortunately, building OpenVPN for Windows on Windows doesn't work well and isn't supported; cross-compiling on Linux is the preferred build method. Our "build_openvpn_windows" function in the current version of build.py is meant to be run on Linux to build Windows, not to be run on Windows.

I think what you need to do is encapsulate these other clones in a "if platform.system() != 'Windows':" block, as it's on Linux and now Darwin we do need them, and then add back running the Windows build on Linux.

minabarret replied:
Me tried to work in this knowledge. Still expect that windows build is failing no linux with mingw at hand currentlly

kevinmkane commented on openvpn/build/build.py:
```diff
@@ -152,131 +217,481 @@ def on_error(func, path, exc_info):
+            git_clone(
+                OPENSSL_OQS_REPO, OPENSSL_OQS_BRANCH,
+                OPENSSL_OQS_REPO_WIN_DIRNAME, OPENSSL_OQS_COMMIT)
+            git_clone(
```
So you can do this, clone OpenSSL-OQS into OPENSSL_OQS_REPO_WIN_DIRNAME, though I'm not sure you need to handle this separately. I think you could always clone into OPENSSL_OQS_REPO_DIRNAME, and then if on Linux just do the build inside it once, or if on Windows, copy it twice for the x86 and x64 builds and then do those builds in there.

kevinmkane commented on openvpn/build/build.py:
```diff
@@ -152,131 +217,481 @@ def on_error(func, path, exc_info):
+parser.add_argument(
+    '--skip-test', action='store_true',
+    help=('skip running make test on build binaries.'),
+    default=False)
```
Thanks for adding this skip test option. The test failures appear to be a problem from upstream, but they don't affect the performance of the PQ algorithms.

kevinmkane commented on openvpn/build/build.py:
```diff
@@ -152,131 +217,481 @@ def on_error(func, path, exc_info):
+parser.add_argument(
+    '--skip-download', action='store_true',
+    help=('skip downloading files, use BUILD_DIR contents.'),
+    default=False)
```
So right now we have two scripts: build.py, which pulls from official sources and builds, and devbuild.py, which uses existing repos to allow developers to test changes on the fly. They're mostly copies of one another with some minor changes.

It looks like with --skip-download it may be possible for a developer to clone their own repos into the correct locations, make local changes, and build with them. Is this correct? If so, I think we can also delete devbuild.py as part of this change.

minabarret replied:
No, its a build option that allows separating the download (network intensive) and the build (cpu-intensive) the --no-clean option may be used for development works, but needs more enhancement in git_restore (eg not unpacking when options.no_clean)
My opinion too is that a separate devbuild.py is overkill, should all be handled in single script

kevinmkane reviewed (changes requested):
Thanks for contributing this work! We're definitely interested in both improving our build script, and support for Darwin is certainly welcome.

I've added some comments based on my testing of your changes on Windows and Linux. I haven't confirmed your script works on Darwin since you've done that, and there are a number of fixes that will be needed to mainly get the Windows build working again. See my comments inline about how we build for Windows; it is a weird two-step process! If you can address these so we have it working on all platforms, we'd be delighted to accept your changes.

minabarret committed:
Enhance build.py

Add info_build will summarise selected command line options
Add BUILD_TARGET_SYSTEM for building for windows in linux
Add notes about building on windows as stated in comments in POST messages. Drop build_openvpn_windows use mingw statements instead.
Die when git_checkout with commit is detected for existing target directory
Add mkdir(STAGE_DIR)

minabarret commented:
Thank you for review and comments. Me updated the pull request with some enhancement.
As i understand it the existence of the python script is used to simplify build and makes special decisions on ssl and openvpn.
What i cannot get in head is why you not used openvpn-build with patches? For what i understand the PQ algorithms will need to be part of openvpn&openssl anyway in future. Then the ssl should crosscompile on linux as well.
The PR is only tested for build on darwin, not linux nor windows

kevinmkane commented:
The PQ algorithms will need to be a part of OpenVPN and OpenSSL in the future, but we're a long way from that. Right now the build process is a little strange. The PQ library currently uses a different build process from the rest of OpenSSL that just doesn't work properly on cross-compiling, which is why we have to undertake this strange process. It may in the future, but it's still very much a work in progress.

I'll take a look at your updated PR on Linux and Windows and see how it works, and get back to you with the results. Thanks again!

kevinmkane commented on openvpn/build/build.py:
```diff
@@ -674,12 +720,25 @@ def post_messages():
+
 mkdir(BUILD_DIR)
 mkdir(IMAGE_DIR)
+mkdir(STAGE_DIR)
```
This works, but we need to also create "x64" and "x86" subdirectories as well. Otherwise we end up with files called "x64" and "x86" that is the contents of one of the binaries. We might consider adding a "/." to the end of the lines where we copy the binaries over to make sure x64 and x86 exist as directories.

minabarret replied:
mh, adding a / to the end seems like a hack, it would be better in I opinion to create the directories prior to the copy (L398/427) as it feels closer to a 'make install'

kevinmkane commented on openvpn/build/build.py:
```diff
@@ -152,131 +281,476 @@ def on_error(func, path, exc_info):
+            die(
+                'Did not find include_dir: {include_dir} directory.'.format(
+                    include_dir=include_dir))
+
```
So in the current official version of this script, the entire stage subtree gets created by the script, but now your changes seem to expect that subtree to expect ahead of time. Can you confirm this is the intended behavior and, if so, why you changed this?

minabarret replied:
when openvpn is build, it expect ssl library to be already compiled and placed in the lib_dir/include_dir location. later the script instruct the compiler (-L{lib_dir}) to use it, not doing the test would cause openvpn to find any installed (ubuntu) ssl instead and not complain. (confirm)

why: me changed from manually creating the stage/subdirs to make install with prefix because its the clean and correct way to join libraries with applications. mkdir/cp instead mimics the intentions of the library author and will most likely miss things and make it harder to understand the staged structure.

kevinmkane commented on openvpn/build/build.py:
```diff
@@ -152,131 +281,476 @@ def on_error(func, path, exc_info):
+    '''
+    Pack the redistributable image for windows
+    '''
+    with chdir(OPENVPN_BUILD_REPO_DIRNAME):
```
I'm getting an odd error at this line when I run: python build.py --target=Windows --skip-test

Traceback (most recent call last):
  File "build.py", line 754, in <module>
    pack_images()
  File "build.py", line 627, in pack_images
    pack_images_windows()
  File "build.py", line 522, in pack_images_windows
    with chdir(OPENVPN_BUILD_REPO_DIRNAME):
  File "/usr/lib/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "build.py", line 120, in chdir
    os.chdir(path)

Before this, it does appear to configure and make in the build/build/openvpn-2.4.4 subtree. Can you investigate?

minabarret replied:
OPENVPN_BUILD_REPO_DIRNAME is a relative name that only exists in SOURCES_DIR after download_remote_repositories. I think the windows build needs love.
but when you got to this point, the build_openvpn was success.
Rewrite pack_images_windows is required when result tree is reviewed

minabarret replied:
the original script has executed build in the openvpn-build and copied the resulting installer. This has changed now as openvpn-build is no longer used (shared build_openvpn) and the openvpn*.exe should be build now as it is done for the other os's.
I know that simple window user needs installer and it is convenient to use the openvpn-build included, but you may need to consider to create a wix/nsis script that does that with the required README (as in linux) and correct custom naming (otherwise real openvpn installation may collide, auto-upgrade whatever magic wus does).
as osx and linux distribute packed files only (not dmg, not deb,rpm,snap), it may be ok for the intended audience to zip the windows files too along with a description to obtain and install tap-drivers, run as admin etc.

when openvpn-build works with patches it can do the installer and all, but as me understand the development is not at a state to do this now.

kevinmkane reviewed (changes requested):
This update seems to have fixed the Linux build, which is great! The Windows build still has some issues, though, which I've noted inline.

If you don't have a Windows system to test on but can test on Linux, we can provide Windows-built OpenSSL binaries to test the Linux build process with. Let me know.

kevinmkane commented:
Okay, I see your points, but this means there's some substantial work beyond what you've done to get the Windows build working again. For us, keeping the Windows build working is more important than adding MacOS, and I don't want to leave your pull request in limbo.

Therefore, I propose we accept your changes into a branch for now, so people who want to build for MacOS can do so, and then we'll open an issue to track getting the Windows build working again in this branch, and then when that's working, we can merge it back to master.

If this is acceptable to you, I'll create a branch, change your pull request to target it, and accept it.
